release/src/router/gettext/gettext-tools/man/msgen.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.24.
.TH MSGEN "1" "May 2005" "GNU gettext-tools 0.14.5" GNU
.SH NAME
msgen \- create English message catalog
.SH SYNOPSIS
.B msgen
[\fIOPTION\fR] \fIINPUTFILE\fR
.SH DESCRIPTION
.\" Add any additional description here
.PP
Creates an English translation catalog.  The input file is the last
created English PO file, or a PO Template file (generally created by
xgettext).  Untranslated entries are assigned a translation that is
identical to the msgid.
.PP
Mandatory arguments to long options are mandatory for short options too.
.SS "Input file location:"
.TP
INPUTFILE
input PO or POT file
.TP
\fB\-D\fR, \fB\-\-directory\fR=\fIDIRECTORY\fR
add DIRECTORY to list for input files search
.PP
If input file is -, standard input is read.
.SS "Output file location:"
.TP
\fB\-o\fR, \fB\-\-output\-file\fR=\fIFILE\fR
write output to specified file
.PP
The results are written to standard output if no output file is specified
or if it is -.
.SS "Input file syntax:"
.TP
\fB\-P\fR, \fB\-\-properties\-input\fR
input file is in Java .properties syntax
.TP
\fB\-\-stringtable\-input\fR
input file is in NeXTstep/GNUstep .strings syntax
.SS "Output details:"
.TP
\fB\-e\fR, \fB\-\-no\-escape\fR
do not use C escapes in output (default)
.TP
\fB\-E\fR, \fB\-\-escape\fR
use C escapes in output, no extended chars
.TP
\fB\-\-force\-po\fR
write PO file even if empty
.TP
\fB\-i\fR, \fB\-\-indent\fR
indented output style
.TP
\fB\-\-no\-location\fR
suppress '#: filename:line' lines
.TP
\fB\-\-add\-location\fR
preserve '#: filename:line' lines (default)
.TP
\fB\-\-strict\fR
strict Uniforum output style
.TP
\fB\-p\fR, \fB\-\-properties\-output\fR
write out a Java .properties file
.TP
\fB\-\-stringtable\-output\fR
write out a NeXTstep/GNUstep .strings file
.TP
\fB\-w\fR, \fB\-\-width\fR=\fINUMBER\fR
set output page width
.TP
\fB\-\-no\-wrap\fR
do not break long message lines, longer than
the output page width, into several lines
.TP
\fB\-s\fR, \fB\-\-sort\-output\fR
generate sorted output
.TP
\fB\-F\fR, \fB\-\-sort\-by\-file\fR
sort output by file location
.SS "Informative output:"
.TP
\fB\-h\fR, \fB\-\-help\fR
display this help and exit
.TP
\fB\-V\fR, \fB\-\-version\fR
output version information and exit
.SH AUTHOR
Written by Bruno Haible.
.SH "REPORTING BUGS"
Report bugs to <bug-gnu-gettext@gnu.org>.
.SH COPYRIGHT
Copyright \(co 2001-2005 Free Software Foundation, Inc.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
The full documentation for
.B msgen
is maintained as a Texinfo manual.  If the
.B info
and
.B msgen
programs are properly installed at your site, the command
.IP
.B info msgen
.PP
should give you access to the complete manual.


CHANGELOG
Tomato USB Mod (ND version)
===========================

build 01 - 12/17/2008
---------------------

Based on the official Tomato 1.23 ND.

Added support for USB 1.1 (OHCI and UHCI) and USB 2.0.

USB support is configurable via Tomato GUI.

USB storage (Ext2, Ext3 and FAT file systems) and USB printing support.

Non-spooling printer server (p910nd 0.92) is included and started
automatically if you enable printing support. Bidirectional copying can
be disabled via GUI if it causes problems with your printer.

Fixed slow running clock problem on Asus WL-520GU. After flashing this
firmware you have to reboot the router at least once for clock to get
fixed.

UPnP is disabled by default.

Added CGI support to HTTP daemon - you can place your cgi scripts into
"/www/ext/cgi-bin" folder, and they will be executed when you access them
via http://<router_ip>/ext/cgi-bin/my_script_name.


build 02 - 12/20/2008
---------------------

Allowed changing USB settings without rebooting a router.

Fixed non-working AIR (WLAN) LED on Asus WL-520GU.

USB support code clean-up.


build 03 - 12/22/2008
---------------------

Cosmetic changes only.


build 04 - 12/26/2008
---------------------

Added FTP server daemon (vsftpd 2.0.7) configurable via GUI.

Updated kernel printer driver to version 0.13 and included some fixes
to the printer driver made by Asus.


build 05 - 12/31/2008 and 07 - 01/03/2009 (UNSTABLE)
------------------------------------------------------

Attempt to update Broadcom Wireless and Ethernet Controller driver to the
version 4.150.10.X (original Tomato included driver is version 4.130.19.0).

Added new setting to "QoS -> Basic Settings" page to select Qdisc
Scheduler (sfq | pfifo).

Auto-mounting for USB storage drives. When this feature is activated, and
USB drive is plugged in, all mountable partitions will be automatically
mounted to a new directories under /mnt. If you want to override default
mount points, mount your partition manually in the satrtup script, i.e.:
"mount -o noatime /dev/discs/disc0/part1 /opt".
If you mount your partition manually, it will not be re-mounted to a
default mount point.

This build, as well as builds 06 and 07 had several issues: using Wireless
Survey and/or changing some settings could crash the router or cause it
to hang.


build 08 - 01/07/2009
---------------------

Since both previous versions with updated WL driver were unstable, rolled
back Broadcom Wireless Controller driver to the old version 4.130.19.0
(original Tomato ND version).

Fixed several bugs introduced in the previous releases.


build 09 - 01/09/2009
---------------------

Added Network File Sharing - Samba 2.0.10.

Added USB Hotplug script - useful for configuring attached USB devices,
i.e. loading firmware into HP1018/1020 or similar printer.

Updated USB drivers to the newer version from kernel 2.4.35/37, applied
some patches to stabilize USB.

Applied patch to allow using routers with up to 128 MB RAM.


build 11 - 01/16/2009
---------------------

Upgraded Broadcom Wireless driver to version 4.150.10.5.

Included SpeedMod by Rodney Chua.

Added functionality to the GUI to unmount (safely remove) USB drives,
reworked the display of attached usb devices.

Fixed "Kernel BUG" error when transfering files to shared USB drives
using Samba.

Added "Regulatory Mode" setting (off|g|h) to "Advanced -> Wireless"
page. This solves long-standing Tomato connectivity problems for Intel
Centrino (2200bg etc) wireless cards.

Allowed up to 100 applications in UPnP.


build 12 - 01/25/2009
---------------------

Bug fixes: cosmetic fixes for GUI, using "Measure Noise Floor" function
on the Device List page should no longer disable wireless.

Improved support for Ext2/Ext3 filesystems (allow expanded inodes,
forward compatibility).

Replaced UPnP daemon with MiniUPnPd.

Restored the gpio utility temporarily removed from previous build.


build 14 - 01/27/2009
---------------------

Minor improvements for FAT filesystem support.

Improved SpeedMod implementation.


build 15 - 01/29/2009
---------------------

Bug fix: corrected firewall rules when WAN access is enabled for FTP.


build 16 - 01/31/2009
---------------------

Fixed error connecting to the FTP server when directory listings are
disabled.

Added support for mounting loopback devices. The loop device module loop.o
is not included but available separately, and can be loaded with insmod
command.

Added option to automatically share all mounted USB partitions.


build 17 - 02/04/2009
---------------------

Patch to resolve read errors on spin-up with some auto spin-down USB
drives (not tested).

Miniupnpd daemon updated to the latest version 20090129. Fixed UPnP
presentation URL when non-standard http port is used.

Fixed LEDs/buttons/GPIO bug introduced with WL driver update (only
affected some router models).

Code clean up, cosmetics.


build 18 - 02/05/2009
---------------------

Removed excessive stripping of shared libraries (fixes problem introduced
in build 17 when using IPP2P in QoS rules).

Added limited kernel-level support for national codepages needed by Samba.


build 19 - 02/10/2009
---------------------

Restored original Tomato setting to show/hide UPnP server in "My Network
Places".

Added scsi-stop command to spin down some USB drives (find a description
at http://wl500g.info/showpost.php?p=47135&postcount=10).

Added GUI settings to require authentication to access Samba shares.

Upgraded Broadcom Wireless driver to version 4.158.4.0.

p910nd printer server is updated to the latest ver. 0.93.


build 20 - 02/24/2009
---------------------

Updated MiniUPnPd to the latest ver. 20090214 and recompiled it with
support for GENA UPnP events enabled.

Updated FTP Server vsftpd to the latest ver. 2.1.0.

Removed non-working ipkg.

Fixed "Unknown Host" Samba error when router is working in Wireless
Bridge or Wireless Client mode.

Fixed SES button and WLAN LED on Asus WL-500gPv1 and Buffalo WBR2-G54
routers.


release/src/router/gettext/gettext-tools/src/msgen.c
/* Creates an English translation catalog.
   Copyright (C) 2001-2005 Free Software Foundation, Inc.
   Written by Bruno Haible <haible@clisp.cons.org>, 2001.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.  */


#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <getopt.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <locale.h>

#include "closeout.h"
#include "dir-list.h"
#include "error.h"
#include "error-progname.h"
#include "progname.h"
#include "relocatable.h"
#include "basename.h"
#include "message.h"
#include "read-po.h"
#include "msgl-english.h"
#include "write-po.h"
#include "exit.h"
#include "gettext.h"

#define _(str) gettext (str)


/* Force output of PO file even if empty.  */
static int force_po;

/* Long options.  */
static const struct option long_options[] =
{
  { "add-location", no_argument, &line_comment, 1 },
  { "directory", required_argument, NULL, 'D' },
  { "escape", no_argument, NULL, 'E' },
  { "force-po", no_argument, &force_po, 1 },
  { "help", no_argument, NULL, 'h' },
  { "indent", no_argument, NULL, 'i' },
  { "no-escape", no_argument, NULL, 'e' },
  { "no-location", no_argument, &line_comment, 0 },
  { "no-wrap", no_argument, NULL, CHAR_MAX + 1 },
  { "output-file", required_argument, NULL, 'o' },
  { "properties-input", no_argument, NULL, 'P' },
  { "properties-output", no_argument, NULL, 'p' },
  { "sort-by-file", no_argument, NULL, 'F' },
  { "sort-output", no_argument, NULL, 's' },
  { "strict", no_argument, NULL, 'S' },
  { "stringtable-input", no_argument, NULL, CHAR_MAX + 2 },
  { "stringtable-output", no_argument, NULL, CHAR_MAX + 3 },
  { "version", no_argument, NULL, 'V' },
  { "width", required_argument, NULL, 'w', },
  { NULL, 0, NULL, 0 }
};


/* Forward declaration of local functions.  */
static void usage (int status)
#if defined __GNUC__ && ((__GNUC__ == 2 && __GNUC_MINOR__ >= 5) || __GNUC__ > 2)
	__attribute__ ((noreturn))
#endif
;

#if defined(__UCLIBC__) && __UCLIBC_MINOR__ == 9 && __UCLIBC_SUBLEVEL__ == 29
void (*error_print_progname) (void) = NULL;
#endif

int
main (int argc, char **argv)
{
  int opt;
  bool do_help;
  bool do_version;
  char *output_file;
  msgdomain_list_ty *result;
  bool sort_by_filepos = false;
  bool sort_by_msgid = false;

  /* Set program name for messages.  */
  set_program_name (argv[0]);
  error_print_progname = maybe_print_progname;

#ifdef HAVE_SETLOCALE
  /* Set locale via LC_ALL.  */
  setlocale (LC_ALL, "");
#endif

  /* Set the text message domain.  */
  bindtextdomain (PACKAGE, relocate (LOCALEDIR));
  textdomain (PACKAGE);

  /* Ensure that write errors on stdout are detected.  */
  atexit (close_stdout);

  /* Set default values for variables.  */
  do_help = false;
  do_version = false;
  output_file = NULL;

  while ((opt = getopt_long (argc, argv, "D:eEFhio:pPsVw:", long_options, NULL))
	 != EOF)
    switch (opt)
      {
      case '\0':		/* Long option.  */
	break;

      case 'D':
	dir_list_append (optarg);
	break;

      case 'e':
	message_print_style_escape (false);
	break;

      case 'E':
	message_print_style_escape (true);
	break;

      case 'F':
	sort_by_filepos = true;
	break;

      case 'h':
	do_help = true;
	break;

      case 'i':
	message_print_style_indent ();
	break;

      case 'o':
	output_file = optarg;
	break;

      case 'p':
	message_print_syntax_properties ();
	break;

      case 'P':
	input_syntax = syntax_properties;
	break;

      case 's':
	sort_by_msgid = true;
	break;

      case 'S':
	message_print_style_uniforum ();
	break;

      case 'V':
	do_version = true;
	break;

      case 'w':
	{
	  int value;
	  char *endp;
	  value = strtol (optarg, &endp, 10);
	  if (endp != optarg)
	    message_page_width_set (value);
	}
	break;

      case CHAR_MAX + 1: /* --no-wrap */
	message_page_width_ignore ();
	break;

      case CHAR_MAX + 2: /* --stringtable-input */
	input_syntax = syntax_stringtable;
	break;

      case CHAR_MAX + 3: /* --stringtable-output */
	message_print_syntax_stringtable ();
	break;

      default:
	usage (EXIT_FAILURE);
	break;
      }

  /* Version information is requested.  */
  if (do_version)
    {
      printf ("%s (GNU %s) %s\n", basename (program_name), PACKAGE, VERSION);
      /* xgettext: no-wrap */
      printf (_("Copyright (C) %s Free Software Foundation, Inc.\n\
This is free software; see the source for copying conditions.  There is NO\n\
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n\
"),
	      "2001-2005");
      printf (_("Written by %s.\n"), "Bruno Haible");
      exit (EXIT_SUCCESS);
    }

  /* Help is requested.  */
  if (do_help)
    usage (EXIT_SUCCESS);

  /* Test whether we have an .po file name as argument.  */
  if (optind >= argc)
    {
      error (EXIT_SUCCESS, 0, _("no input file given"));
      usage (EXIT_FAILURE);
    }
  if (optind + 1 != argc)
    {
      error (EXIT_SUCCESS, 0, _("exactly one input file required"));
      usage (EXIT_FAILURE);
    }

  /* Verify selected options.  */
  if (!line_comment && sort_by_filepos)
    error (EXIT_FAILURE, 0, _("%s and %s are mutually exclusive"),
	   "--no-location", "--sort-by-file");

  if (sort_by_msgid && sort_by_filepos)
    error (EXIT_FAILURE, 0, _("%s and %s are mutually exclusive"),
	   "--sort-output", "--sort-by-file");

  /* Read input file and add English translations.  */
  result = msgdomain_list_english (read_po_file (argv[optind]));

  /* Sort the results.  */
  if (sort_by_filepos)
    msgdomain_list_sort_by_filepos (result);
  else if (sort_by_msgid)
    msgdomain_list_sort_by_msgid (result);

  /* Write the merged message list out.  */
  msgdomain_list_print (result, output_file, force_po, false);

  exit (EXIT_SUCCESS);
}


/* Display usage information and exit.  */
static void
usage (int status)
{
  if (status != EXIT_SUCCESS)
    fprintf (stderr, _("Try `%s --help' for more information.\n"),
	     program_name);
  else
    {
      printf (_("\
Usage: %s [OPTION] INPUTFILE\n\
"), program_name);
      printf ("\n");
      /* xgettext: no-wrap */
      printf (_("\
Creates an English translation catalog.  The input file is the last\n\
created English PO file, or a PO Template file (generally created by\n\
xgettext).  Untranslated entries are assigned a translation that is\n\
identical to the msgid.\n\
"));
      printf ("\n");
      printf (_("\
Mandatory arguments to long options are mandatory for short options too.\n"));
      printf ("\n");
      printf (_("\
Input file location:\n"));
      printf (_("\
  INPUTFILE                   input PO or POT file\n"));
      printf (_("\
  -D, --directory=DIRECTORY   add DIRECTORY to list for input files search\n"));
      printf (_("\
If input file is -, standard input is read.\n"));
      printf ("\n");
      printf (_("\
Output file location:\n"));
      printf (_("\
  -o, --output-file=FILE      write output to specified file\n"));
      printf (_("\
The results are written to standard output if no output file is specified\n\
or if it is -.\n"));
      printf ("\n");
      printf (_("\
Input file syntax:\n"));
      printf (_("\
  -P, --properties-input      input file is in Java .properties syntax\n"));
      printf (_("\
      --stringtable-input     input file is in NeXTstep/GNUstep .strings syntax\n"));
      printf ("\n");
      printf (_("\
Output details:\n"));
      printf (_("\
  -e, --no-escape             do not use C escapes in output (default)\n"));
      printf (_("\
  -E, --escape                use C escapes in output, no extended chars\n"));
      printf (_("\
      --force-po              write PO file even if empty\n"));
      printf (_("\
  -i, --indent                indented output style\n"));
      printf (_("\
      --no-location           suppress '#: filename:line' lines\n"));
      printf (_("\
      --add-location          preserve '#: filename:line' lines (default)\n"));
      printf (_("\
      --strict                strict Uniforum output style\n"));
      printf (_("\
  -p, --properties-output     write out a Java .properties file\n"));
      printf (_("\
      --stringtable-output    write out a NeXTstep/GNUstep .strings file\n"));
      printf (_("\
  -w, --width=NUMBER          set output page width\n"));
      printf (_("\
      --no-wrap               do not break long message lines, longer than\n\
                              the output page width, into several lines\n"));
      printf (_("\
  -s, --sort-output           generate sorted output\n"));
      printf (_("\
  -F, --sort-by-file          sort output by file location\n"));
      printf ("\n");
      printf (_("\
Informative output:\n"));
      printf (_("\
  -h, --help                  display this help and exit\n"));
      printf (_("\
  -V, --version               output version information and exit\n"));
      printf ("\n");
      fputs (_("Report bugs to <bug-gnu-gettext@gnu.org>.\n"),
	     stdout);
    }

  exit (status);
}
